### DISCUSSION TITLE
Relationship with https://github.com/containers/bootc ?

### DISCUSSION THREAD

**Discussion by cgwalters** [foresterorg/forester#7] [General]
I only just saw this project, but I wanted to ask about the relationship you see with the [bootc project](https://github.com/containers/bootc).

It looks to me like the systems here are always run as a "Live ISO"?  Presumably with an optional stateful partition?  In other words, in-place updates aren't scoped in; other than supporting `dnf update` implicitly too?  But with any updates applied there discarded on reboot?

**Comment by lzap** [comment]
Hello Colin,

what a coincidence, I spent an hour in the morning learning more about bootc and related projects. From what I understand, both projects complement each other greatly, let me explain.

Forester is nothing else than a HTTP endpoint that provides required functionality to perform unattended Anaconda installations via `liveimg` statement. This feature was added back in EL 7.2 for oVirt (RHEL-H deployments) and is now utilized for R4E as well (Anaconda can install R4E, I know there is also coreos-installer too). Because it also works for standard RHEL, the idea is to provide the same workflow for both RHEL and R4E.

I would like to know where you got the impression of that this is all about LiveCD/ISO. This looks like the project home page does fail quite a bit in explaining what the project does :-) After a RHEL/Fedora system in installed via `liveimg`, it is a full installation without any limitations. This either works in a traditional fashion (RPM) or via an ostree (Edge).

Main features of the project are:

* Host `shim.efi` and `grubxxx.efi` bootloader. Forester currently uses both files from the very fist uploaded image, it can be configured to use a different image. No BIOS support, it only works with UEFI.
* Host Anaconda stage2 files and images (tarballs). These files are extracted as users upload their images into Forester and served over HTTP.
* Renders grub template, which depending on the MAC address reported by grub and loads kernel/initrd over HTTP UEFI Boot. PXE (over EFI) is not supported.
* Anaconda loads in and asks for the kickstart which is rendered depending on the MAC address.
* There is an additional discovery process for hosts which are not yet in the inventory which is implemented as a `%pre` Anaconda script.
* If a host is configured for provisioning, kickstart is returned with the `liveimg` statement pointing to the tarball.
* Power management over Redfish (or libvirt) API so users get full end to end experience using Forester API/CLI.

Feel free to drop me an invite if you want to talk more.

**Reply by cgwalters** [reply]
I didn't even know about `liveimg.`  It's like anaconda embedding coreos-installer it seems?

> I would like to know where you got the impression of that this is all about LiveCD/ISO. 

Mainly because the kickstart verb is called `liveimg` :smile: 

But yes it's really more like `diskimg` it seems.  So the way I think of is this is that the [proposed `bootc` verb](https://github.com/rhinstaller/anaconda/discussions/5197) (and existing `ostreecontainer` verb) replace `liveimg` in this use case.

Basically instead of managing disk images stored on a http server with manual sha256 checksums, one stores OS content in a registry.

We're having a parallel discussion in another place as to whether it makes sense to support both "bootc/ostree style" installs and dnf-managed installs with a single container image.

**Reply by lzap** [reply]
Oh I just also learned about the new [ostreecontainer](https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#ostreecontainer) command. Okay, that changes a lot of things. I somehow thought that the bootc project uses `liveimg` statement, now I see.

In this case, I think it would make sense to redesign Forester to be `ostreecontainer` native. This also means substantial changes of how Forester work, there is no reason to upload images as, if I understand correctly, `ostreecontainer` connect directly to a registry and downloads/installs "on the fly".

> But yes it's really more like diskimg it seems. So the way I think of is this is that the https://github.com/rhinstaller/anaconda/discussions/5197 (and existing ostreecontainer verb) replace liveimg in this use case.

Ah, now I understand. Okay, interesting, we need to talk.

**Reply by cgwalters** [reply]
> In this case, I think it would make sense to redesign Forester to be ostreecontainer native. 

Yes I think that'd be cool :smile:

**Reply by cgwalters** [reply]
> if I understand correctly, ostreecontainer connect directly to a registry and downloads/installs "on the fly".

Exactly.  Not just that, but "day 2" one types `bootc update` for in-place transactional updates to new images pushed to the registry.

**Reply by lzap** [reply]
Ok I will look into implementing support for `ostreecontainer` in Forester. Thanks for the info!
